vere/dawn.c

```diff
@@ -0,0 +1,444 @@
+    }
+    else if ( c3__pawn == rank ) {
+      //  irrelevant, just bunt +hull
+      //
+      hul = u3v_wish("*hull:constitution:ethe");
+    }
+    else {
+      u3_noun oct;
+
+      if ( c3__earl == rank ) {
+        u3_noun seg = u3do("^sein:title", u3k(ship));
+        u3_noun ges = u3dc("scot", 'p', u3k(seg));
+        c3_c* seg_c = u3r_string(ges);
+
+        fprintf(stderr, "boot: retrieving %s's public keys (for %s)\r\n",
+                                            seg_c, u3_Host.ops_u.who_c);
+        oct = u3dc("hull:give:dawn", u3k(bok), u3k(seg));
+
+        free(seg_c);
+        u3z(seg); u3z(ges);
+      }
+      else {
+        fprintf(stderr, "boot: retrieving %s's public keys\r\n",
+                                           u3_Host.ops_u.who_c);
+        oct = u3dc("hull:give:dawn", u3k(bok), u3k(ship));
+      }
+
+      u3_noun luh = _dawn_eth_rpc(url_c, u3k(oct));
+      hul = u3dc("hull:take:dawn", u3k(ship), u3k(luh));
+
+      u3z(oct); u3z(luh);
+    }
+
+    //  +live:dawn: network state
+    //  XX actually make request
+    //
+    u3_noun liv = u3_nul;
+    // u3_noun liv = _dawn_get_json(parent, /some/url)
+
+    fprintf(stderr, "boot: verifying keys\r\n");
+
+    //  (each sponsor=(unit ship) error=@tas)
+    //
+    u3_noun sas = u3dt("veri:dawn", u3k(seed), u3k(hul), u3k(liv));
+
+    if ( c3n == u3h(sas) ) {
+      // bails, won't return
+      _dawn_fail(ship, rank, sas);
+      return u3_none;
+    }
+
+    //  (unit ship): sponsor
+    //  produced by +veri:dawn to avoid coupling to +hull structure
+    //  XX reconsider
+    //
+    pon = u3k(u3t(sas));
+
+    u3z(hul); u3z(liv); u3z(sas);
+  }
+
+  //  (map ship [=life =pass]): galaxy table
+  //
+  if ( c3y == u3_Host.ops_u.etn ) {
+    zar = u3do("czar:snap:dawn", u3k(u3t(sap)));
+  }
+  else {
+    fprintf(stderr, "boot: retrieving galaxy table\r\n");
+
+    u3_noun oct = u3do("czar:give:dawn", u3k(bok));
+    u3_noun raz = _dawn_eth_rpc(url_c, u3k(oct));
+    zar = u3do("czar:take:dawn", u3k(raz));
+
+    u3z(oct); u3z(raz);
+  }
+
+  //  (list turf): ames domains
+  //
+  if ( 0 != u3_Host.ops_u.dns_c ) {
+    tuf = _dawn_turf(u3_Host.ops_u.dns_c);
+  }
+  else if ( c3y == u3_Host.ops_u.etn ) {
+    tuf = u3do("turf:snap:dawn", u3k(u3t(sap)));
+  }
+  else {
+    fprintf(stderr, "boot: retrieving network domains\r\n");
+
+    u3_noun oct = u3do("turf:give:dawn", u3k(bok));
+    u3_noun fut = _dawn_eth_rpc(url_c, u3k(oct));
+    tuf = u3do("turf:take:dawn", u3k(fut));
+
+    u3z(oct); u3z(fut);
+  }
+
+  u3z(rank);
+
+  //  [%dawn seed sponsor galaxies domains block eth-url snap]
+  //
+  return u3nc(c3__dawn, u3nq(seed, pon, zar, u3nq(tuf, bok, url, sap)));
+}
+
+/* u3_dawn_come(): mine a comet under star (unit)
+*/
+u3_noun
+u3_dawn_come(u3_noun star)
+{
+  u3_noun seed;
+
+  if ( u3_nul == star ) {
+    //  XX ~marzod hardcoded
+    //  choose from list, at random, &c
+    //
+    star = 256;
```

Comment:
Also not related to this patch, but we have a weird tendency to use decimal in lots of places where the hexadecimal would be clearer and would reduce the chances of an off-by-one error.